sdcm/cluster.py

```diff
@@ -2812,20 +2812,20 @@ class FlakyRetryPolicy(RetryPolicy):
     A retry policy that retries 5 times
     """
 
-    def _retry_message(self, msg, *args, **kwargs):  # pylint: disable=unused-argument,arguments-differ
+    def _retry_message(self, msg, /, *args, **kwargs):  # pylint: disable=unused-argument,arguments-differ
         if kwargs['retry_num'] < 5:
             LOGGER.debug("%s. Attempt #%s", msg, str(kwargs['retry_num']))
             return self.RETRY, None
         return self.RETHROW, None
 
     def on_read_timeout(self, *args, **kwargs):
-        return self._retry_message(msg="Retrying read after timeout", *args, **kwargs)
+        return self._retry_message("Retrying read after timeout", *args, **kwargs)
 
     def on_write_timeout(self, *args, **kwargs):
-        return self._retry_message(msg="Retrying write after timeout", *args, **kwargs)
+        return self._retry_message("Retrying write after timeout", *args, **kwargs)
 
     def on_unavailable(self, *args, **kwargs):
-        return self._retry_message(msg="Retrying request after UE", *args, **kwargs)
+        return self._retry_message("Retrying request after UE", *args, **kwargs)
```

Comment:
Too much of a black magic for my taste here. It's really hard to understand why things done in this way.

Do you find that it's unnecessary to pass all arguments to `_retry_message`?

I can suggest to make following changes:

```python
def _retry_message(self, msg, retry_num):
    ...

def on_read_timeout(self, *args, retry_num, **kwargs):
    return self._retry_message(msg="...", retry_num=retry_num)

...
```

What do you think about it?

The popping of `retry_num` out of `**kwargs` is based on the same assumption as in the original code that it has been passed as a keyword argument.

btw, the root of cause is how Python driver calls these handlers:

```python
    def on_read_timeout(self, query, consistency, required_responses,
                        received_responses, data_retrieved, retry_num):

...

                if isinstance(response, ReadTimeoutErrorMessage):
                    if self._metrics is not None:
                        self._metrics.on_read_timeout()
                    retry = retry_policy.on_read_timeout(
                        self.query, retry_num=self._query_retries, **response.info)
                elif isinstance(response, WriteTimeoutErrorMessage):
                    ...
```

Reply:
agree. Better to eliminate `*args, **kwargs` from the `_retry_message`. 
And also from `on_read_timeout` since you don't know how they call `on_read_timeout` in all places in the driver (using positional or kw arguments), so it can be possible a situation when you will get `retry_num` None.
I think the best way is :

```python

   def on_read_timeout(self, query, consistency, required_responses,
                        received_responses, data_retrieved, retry_num):    # pylint: disable=unused-argument
        return self._retry_message(msg="...", retry_num=retry_num)
```
and define `_retry_message` as Evgeniy suggested.